### PR TITLE
service discovery modified

### PR DESCRIPTION
**Pull request by kumarabd**
Signed-off-by: kumarabd <abishekkumar92@gmail.com>

**Description**

Discovers service endpoint both internal and external.
First looks for load balancer in case of external service endpoint, and then nodeport.
**Notes for Reviewers**

**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by kumarabd** [comment]
@leecalcote @utkarsh-pro @mgfeller breaking change in the GetServiceEndpoint function. Hope we remember this when bumping versions.

File: utils/kubernetes/service.go [hunk 1]
```diff
@@ -2,109 +2,74 @@ package kubernetes
...
-	return nil
-}
 
-// extractClusterIPEndpoint extracts clusterIP based endpoint, if any.
-// It returns the nil if no valid endpoint is extracted
-func extractClusterIPEndpoint(svc *coreV1.Service) *utils.Endpoint {
-	ports := svc.Spec.Ports
-	clusterIP := svc.Spec.ClusterIP
+	endpoint.External = &utils.HostPort{
+		Address: "localhost",
```
Comment (leecalcote) [inline utils/kubernetes/service.go:58]: Will "localhost" always be the correct reference?

**Review by leecalcote — APPROVED** [review]

File: utils/kubernetes/service.go (hunk 1 above)
Comment (kumarabd) [reply utils/kubernetes/service.go:58]: its the default value.

**Comment by mgfeller** [comment]
> @leecalcote @utkarsh-pro @mgfeller breaking change in the GetServiceEndpoint function. Hope we remember this when bumping versions.

we have tests to discover that, right? :smile:

File: utils/kubernetes/service.go [hunk 2]
```diff
@@ -2,109 +2,74 @@ package kubernetes
...
+	Namespace    string
+	PortSelector string
+	PortNumber   *int32
+	APIServerURL string
+}
 
-	// Try nodeport endpoint
-	nodes, err := client.KubeClient.CoreV1().Nodes().List(ctx, v1.ListOptions{})
+// GetServiceEndpoint returns the endpoint for the given service
+func GetServiceEndpoint(ctx context.Context, client *kubernetes.Clientset, opts *ServiceOptions) (*utils.Endpoint, error) {
```
Comment (mgfeller) [inline utils/kubernetes/service.go:22]: would be nice with some more explanation.

File: utils/kubernetes/service.go [hunk 3]
```diff
@@ -2,109 +2,74 @@ package kubernetes
...
-	// Try loadbalancer endpoint
-	if endpoint := extractLoadBalancerEndpoint(svc); endpoint != nil {
-		return endpoint, nil
-	}
+type ServiceOptions struct {
+	Name         string
+	Namespace    string
+	PortSelector string
+	PortNumber   *int32
+	APIServerURL string
```
Comment (mgfeller) [inline utils/kubernetes/service.go:18]: what's that? would be nice with a comment :smile:

File: utils/kubernetes/service.go [hunk 4]
```diff
@@ -2,109 +2,74 @@ package kubernetes
...
-	}
-
-	// Try loadbalancer endpoint
-	if endpoint := extractLoadBalancerEndpoint(svc); endpoint != nil {
-		return endpoint, nil
-	}
+type ServiceOptions struct {
+	Name         string
+	Namespace    string
+	PortSelector string
```
Comment (mgfeller) [inline utils/kubernetes/service.go:16]: would be appreciated with a comment explaining that, I think.

**Review by mgfeller — COMMENTED** [review]
> @leecalcote @utkarsh-pro @mgfeller breaking change in the GetServiceEndpoint function. Hope we remember this when bumping versions.

what's the breaking change? would be nice to know :smile:

File: utils/kubernetes/service.go (hunk 2 above)
Comment (kumarabd) [reply utils/kubernetes/service.go:22]: @mgfeller  The function definition has been changed, and the output struct is also different. thus breaking change.

File: utils/kubernetes/service.go (hunk 3 above)
Comment (kumarabd) [reply utils/kubernetes/service.go:18]: Ah yes okay adding them

**Comment by kumarabd** [comment]
@mgfeller added some comments

File: utils/kubernetes/service.go (hunk 2 above)
Comment (mgfeller) [reply utils/kubernetes/service.go:22]: @kumarabd ah ok thanks, but the functionality remains the same?

Comment (kumarabd) [reply utils/kubernetes/service.go:22]: yes @mgfeller

**Pull request merged**